fem/fespace.cpp

```diff
@@ -2881,238 +3002,1215 @@ void ElementRestriction::MultTranspose(const Vector& x, Vector& y) const
-}
-
-template<const int T_VDIM, const int T_ND, const int T_NQ>
-void QuadratureInterpolator::Eval2D(
-   const int NE,
-   const int vdim,
-   const DofToQuad &maps,
-   const Vector &e_vec,
-   Vector &q_val,
-   Vector &q_der,
-   Vector &q_det,
-   const int eval_flags)
+void H1FaceRestriction::GetFaceDofs(const int dim, const int face_id,
+                                    const int dof1d, Array<int> &faceMap)
 {
-   const int nd = maps.ndof;
-   const int nq = maps.nqpt;
-   const int ND = T_ND ? T_ND : nd;
-   const int NQ = T_NQ ? T_NQ : nq;
-   const int VDIM = T_VDIM ? T_VDIM : vdim;
-   MFEM_VERIFY(ND <= MAX_ND2D, "");
-   MFEM_VERIFY(NQ <= MAX_NQ2D, "");
-   MFEM_VERIFY(VDIM == 2 || !(eval_flags & DETERMINANTS), "");
-   auto B = Reshape(maps.B.Read(), NQ, ND);
-   auto G = Reshape(maps.G.Read(), NQ, 2, ND);
-   auto E = Reshape(e_vec.Read(), ND, VDIM, NE);
-   auto val = Reshape(q_val.Write(), NQ, VDIM, NE);
-   auto der = Reshape(q_der.Write(), NQ, VDIM, 2, NE);
-   auto det = Reshape(q_det.Write(), NQ, NE);
-   MFEM_FORALL(e, NE,
+   switch (dim)
    {
-      const int ND = T_ND ? T_ND : nd;
-      const int NQ = T_NQ ? T_NQ : nq;
-      const int VDIM = T_VDIM ? T_VDIM : vdim;
-      constexpr int max_ND = T_ND ? T_ND : MAX_ND2D;
-      constexpr int max_VDIM = T_VDIM ? T_VDIM : MAX_VDIM2D;
-      double s_E[max_VDIM*max_ND];
-      for (int d = 0; d < ND; d++)
-      {
-         for (int c = 0; c < VDIM; c++)
+      case 1:
+         switch (face_id)
          {
-            s_E[c+d*VDIM] = E(d,c,e);
+            case 0://WEST
+               faceMap[0] = 0;
+               break;
+            case 1://EAST
+               faceMap[0] = dof1d-1;
+               break;
          }
-      }
-      for (int q = 0; q < NQ; ++q)
-      {
-         if (eval_flags & VALUES)
+         break;
+      case 2:
+         switch (face_id)
          {
-            double ed[max_VDIM];
-            for (int c = 0; c < VDIM; c++) { ed[c] = 0.0; }
-            for (int d = 0; d < ND; ++d)
-            {
-               const double b = B(q,d);
-               for (int c = 0; c < VDIM; c++) { ed[c] += b*s_E[c+d*VDIM]; }
-            }
-            for (int c = 0; c < VDIM; c++) { val(q,c,e) = ed[c]; }
+            case 0://SOUTH
+               for (int i = 0; i < dof1d; ++i)
+               {
+                  faceMap[i] = i;
+               }
+               break;
+            case 1://EAST
+               for (int i = 0; i < dof1d; ++i)
+               {
+                  faceMap[i] = dof1d-1 + i*dof1d;
+               }
+               break;
+            case 2://NORTH
+               for (int i = 0; i < dof1d; ++i)
+               {
+                  faceMap[i] = (dof1d-1)*dof1d + dof1d-1 - i;
+               }
+               break;
+            case 3://WEST
+               for (int i = 0; i < dof1d; ++i)
+               {
+                  faceMap[i] = (dof1d-1)*dof1d - i*dof1d;
+               }
+               break;
```

Comment:
This should be made lexicographic to match with the 3D implementation

Reply:
done.